docs/root/configuration/http_conn_man/headers.rst

```diff
@@ -166,17 +165,88 @@ accurate. The source IP address of the immediate downstream node's connection to
 trusted. XFF *sometimes* can be trusted. Malicious clients can forge XFF, but the last
 address in XFF can be trusted if it was put there by a trusted proxy.
 
-Envoy's rules for determining the trusted client address are:
+Envoy's default rules for determining the trusted client address (*before* appending anything
+to XFF) are:
 
 * If *use_remote_address* is false and an XFF containing at least one IP address is
   present in the request, the trusted client address is the *last* (rightmost) IP address in XFF.
 * Otherwise, the trusted client address is the source IP address of the immediate downstream
   node's connection to Envoy.
 
+In an environment where there are one or more trusted proxies in front of an edge
+Envoy instance, the *xff_num_trusted_hops* configuration option can be used to trust
+additional addresses from XFF:
+
+* If *use_remote_address* is false and *xff_num_trusted_hops* is set to a value *N* that is
+  greater than zero, the trusted client address is the *N+1* th address from the right end
+  of XFF.
+* If *use_remote_address* is true and *xff_num_trusted_hops* is set to a value *N* that is
+  greater than zero, the trusted client address is the *N* th address from the right end
+  of XFF.
+
 Envoy uses the trusted client address contents to determine whether a request originated
 externally or internally. This influences whether the
 :ref:`config_http_conn_man_headers_x-envoy-internal` header is set.
 
+Example 1: Envoy as edge proxy, without a trusted proxy in front of it
+    Settings:
+      | use_remote_address = true
+      | xff_num_trusted_hops = 0
+
+    Request details:
+      | Downstream IP address = 192.0.2.5
+      | XFF = "198.51.100.128, 198.51.100.10, 198.51.100.1"
+
+    Result:
+      | Trusted client address = 192.0.2.5 (XFF is ignored)
+      | X-Envoy-External-Address is set to 192.0.2.5
+      | XFF is changed to "198.51.100.128, 198.51.100.10, 198.51.100.1, 192.0.2.5"
+      | Request type = external
+
+Example 2: Envoy as internal proxy, with the Envoy edge proxy from Example 1 in front of it
+    Settings:
+      | use_remote_address = false
+      | xff_num_trusted_hops = 0
+
+    Request details:
+      | Downstream IP address = 10.11.12.13 (address of the Envoy edge proxy)
+      | XFF = "198.51.100.128, 198.51.100.10, 198.51.100.1, 192.0.2.5"
+
+    Result:
+      | Trusted client address = 192.0.2.5 (last address in XFF is trusted)
+      | X-Envoy-External-Address is not modified
+      | Request type = internal
```

Comment:
This is an internal request, but in this case x-envoy-internal will be set to false, because XFF has multiple addresses in it. I realize this is confusing, potentially wrong, etc., but it's the way it is. Not sure how you want to discuss in docs, but I think actually saying what x-envoy-internal would be in each case is probably worth it.

Reply:
Thanks for catching that! I'll fix the example and add x-envoy-internal.